src/main/java/build/buildfarm/server/ExecutionService.java

```diff
@@ -29,6 +32,58 @@ public ExecutionService(Instances instances) {
     this.instances = instances;
   }
 
+  private Predicate<Operation> createWatcher(StreamObserver<Operation> responseObserver) {
+    return (operation) -> {
+      if (Context.current().isCancelled()) {
+        return false;
+      }
+      try {
+        if (operation != null) {
+          responseObserver.onNext(operation);
+        }
+        if (operation == null || operation.getDone()) {
```

Comment:
if you move the operation == null condition into line 37, then you don't need to check for it in the rest of the function and the code gets slightly simpler.

Reply:
But in the null operation case, I still need to complete the responseObserver stream. Moving it out of the try block here will also prevent coverage by the grpc induced cancellation/expiration handlers present in the catches.

Reply:
Oh, sorry, you're right! Disregard.